### PR TITLE
Locksmith route to migrate lock

### PR DESCRIPTION
**Pull request by clemsos**
# Description

This creates a route in locksmith that can be called to replicate/migrate a lock to a new location.

~~NB: need to be rebase once #7821 is merged~~ done

How it works 

```
$ curl -X POST http://localhost:8080/lock/$LOCK_ADDRESS/migrate                                                                                                                             
{"lockAddress":"0x98E063A6b0BC3dA2a87eAE801deD861D41a0b494","newLockAddress":"0x8388146eF3b9d44a1b1455C4247438B43401E893"}%
```

A record will be added in pg `LockMigrations` table

```
graph-node=# SELECT * FROM "LockMigrations" WHERE id = 51;
-[ RECORD 1 ]--+-----------------------------------------------------------------------------------------------------------------------------------------------
id             | 51
lockAddress    | 0x98E063A6b0BC3dA2a87eAE801deD861D41a0b494
newLockAddress | 0x8388146eF3b9d44a1b1455C4247438B43401E893
chain          | 31337
logs           |                                                                                                                                               +
               | CLONE LOCK > cloning 0x98E063A6b0BC3dA2a87eAE801deD861D41a0b494 on localhost...                                                               +
               | CLONE LOCK > deployed to : 0x8388146eF3b9d44a1b1455C4247438B43401E893 (tx: 0x4d6eb2e53e846bce6c1c94460aaf0ee151289ac1ffbb48db2e97a34a25f48432)+
               | CLONE LOCK > add key owners...                                                                                                                +
               | CLONE LOCK > 10 keys transferred, 10 key managers changed                                                                                     +
               | CLONE LOCK > fetching managers...                                                                                                             +
               | request to http://graph-node:8000/ failed, reason: getaddrinfo ENOTFOUND graph-node                                                           +
               | LOCK CLONE > 0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266 set as beneficiary.                                                                   +
               | LOCK CLONE > Symbol updated to 'KEY'.                                                                                                         +
               | LOCK CLONE > Removed Unlock as lock manager.
migrated       | t
createdAt      | 2021-12-02 16:36:54.733+00
updatedAt      | 2021-12-02 16:37:01.761+00
```

On each attempt a record is created with "migrated" flag set to false. If the migration fails, logs are stored in the `logs` field of the record.

### CODE REVIEW

**Comment by clemsos** [comment]
There are still a few TODOs in the code. Will be clearer once rebased

File: locksmith/src/controllers/lockController.js [hunk 1]
```diff
@@ -36,6 +42,57 @@ const lockGet = async (req, res) => {
...
+})
+
+const lockMigrate = async (req, res) => {
+  const { lockAddress } = req.params
+  const unlockVersion = req.query.unlockVersion || 9
+  const chainId = req.chain
+
+  // TODO: make sure lock exists
+  // const databaseLock = await getLockByAddress(lockAddress)
+  // if (!databaseLock) res.send(404)
```
Comment (clemsos) [inline locksmith/src/controllers/lockController.js:57]: I wasnt sure how to handle errors on locksmith API. Should we just return an error code? or is there sth like a formatted error JSON ?

File: locksmith/src/controllers/lockController.js [hunk 2]
```diff
@@ -36,6 +42,57 @@ const lockGet = async (req, res) => {
   res.json(baseTokenData)
 }
 
+migrateLogEvent.on('migrateLock', async ({ recordId, msg }) => {
+  console.log('migrateLogEvent', recordId, msg)
```
Comment (clemsos) [inline locksmith/src/controllers/lockController.js:46]: using an event to make sure we keep the logs are they arrive

File: locksmith/src/utils/lockMigrate.ts [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+}
+
+class MigrateLogEventEmitter extends EventEmitter {}
+
+export const migrateLogEvent = new MigrateLogEventEmitter()
+
+// TODO: how to pass this safely?
+const mnemonic =
+  process.env.WALLET_MNEMONIC ||
+  'test test test test test test test test test test test junk'
```
Comment (clemsos) [inline locksmith/src/utils/lockMigrate.ts:22]: this remains a challenge as we need to deploy from the server...

**Comment by clemsos** [comment]
this may require #7895 to be merged here to work (so it gets latest serializer)

File: locksmith/src/controllers/lockController.js (hunk 1 above)
Comment (julien51) [reply locksmith/src/controllers/lockController.js:57]: We do not have a standard for now. So I think `res.send(404, "Missing lock")` will do it for now

File: locksmith/src/controllers/lockController.js [hunk 4]
```diff
@@ -36,6 +42,57 @@ const lockGet = async (req, res) => {
...
+  const chainId = req.chain
+
+  // TODO: make sure lock exists
+  // const databaseLock = await getLockByAddress(lockAddress)
+  // if (!databaseLock) res.send(404)
+
+  // record the migration in db
+  const dbRecord = await LockMigrations.create({
+    lockAddress,
+    // initiatedBy: msg.sender // TODO: get that info somewhere?
```
Comment (julien51) [inline locksmith/src/controllers/lockController.js:62]: Initially I was thinking we should ask the user to sign a message  but I don't think this is required and I worry this adds complexity?

File: locksmith/migrations/20211202121847-create-lock-migrations.js [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+        primaryKey: true,
+        type: Sequelize.INTEGER,
+      },
+      lockAddress: {
+        type: Sequelize.STRING,
+      },
+      newLockAddress: {
+        type: Sequelize.STRING,
+      },
+      chain: {
```
Comment (julien51) [inline locksmith/migrations/20211202121847-create-lock-migrations.js:16]: Bon je me dis que ca pourrait etre cool plus tard de migrer d'une chaine a l'autre? On rajoute un `newLockChain` ?

File: locksmith/src/controllers/lockController.js [hunk 6]
```diff
@@ -36,6 +42,57 @@ const lockGet = async (req, res) => {
...
 
+migrateLogEvent.on('migrateLock', async ({ recordId, msg }) => {
+  console.log('migrateLogEvent', recordId, msg)
+  await updateLockMigrationsLog(recordId, msg)
+})
+
+const lockMigrate = async (req, res) => {
+  const { lockAddress } = req.params
+  const unlockVersion = req.query.unlockVersion || 9
+  const chainId = req.chain
```
Comment (julien51) [inline locksmith/src/controllers/lockController.js:53]: Ok mettons une limit sur les chain id juste 100 ou 137 pour l'instant?

File: locksmith/src/controllers/lockController.js (hunk 4 above)
Comment (julien51) [reply locksmith/src/controllers/lockController.js:62]: As you noted I think good enough to use a query string populated by front-end?

File: locksmith/src/utils/lockMigrate.ts (hunk 3 above)
Comment (julien51) [reply locksmith/src/utils/lockMigrate.ts:22]: Let's use the same credentials as the key purchaser. we already have it setup.

**Review by julien51 — COMMENTED** [review]
ok looking good for now! I need to spend a bit more time on it tonight

File: locksmith/src/controllers/lockController.js (hunk 4 above)
Comment (clemsos) [reply locksmith/src/controllers/lockController.js:62]: yep ok for using URL param

**Comment by clemsos** [comment]
> this may require #7895 to be merged here to work (so it gets latest serializer)

merged this PR here so we can have tests passing. Can rebase once it is merged

File: locksmith/migrations/20211202121847-create-lock-migrations.js (hunk 5 above)
Comment (clemsos) [reply locksmith/migrations/20211202121847-create-lock-migrations.js:16]: let's wait to have an implementation to add a migration? With sequelize its quite straightforward

File: locksmith/src/controllers/lockController.js (hunk 6 above)
Comment (clemsos) [reply locksmith/src/controllers/lockController.js:54]: ok added

**Comment by clemsos** [comment]
> > this may require #7895 to be merged here to work (so it gets latest serializer)
> 
> merged this PR here so we can have tests passing. Can rebase once it is merged

ok rebased

**Review by julien51 — APPROVED** [review]

File: locksmith/src/utils/lockMigrate.ts [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+      })
+    }
+  } catch (error) {
+    migrateLogEvent.emit('migrateLock', {
+      recordId,
+      msg: `LOCK CLONE > baseTokenURI not set, using default. (ex. '${tokenURISample}').`,
+    })
+  }
+
+  if (symbol != 'UDT') {
```
Comment (julien51) [inline locksmith/src/utils/lockMigrate.ts:232]:
```suggestion
  if (symbol != 'KEY') {
```

UDT is the symbole of the Governance token. The symbol for the NFT is`KEY` :)

File: locksmith/src/utils/lockMigrate.ts [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+  // update the beneficiary
+  const txBenef = await newLock.updateBeneficiary(beneficiary)
+  await txBenef.wait()
+  migrateLogEvent.emit('migrateLock', {
+    recordId,
+    msg: `LOCK CLONE > ${beneficiary} set as beneficiary.`,
+  })
+
+  // update metadata
+  if (symbol != 'UDT') {
```
Comment (julien51) [inline locksmith/src/utils/lockMigrate.ts:203]:
```suggestion
  if (symbol != 'KEY') {
```

**Comment by clemsos** [comment]
- [] prevent duplicate migration at route level
- [] make sure we transfer the lock (set lock manager) at the very end of the migration process
- [] url param to force migration

**Comment by clemsos** [comment]
I have started to move the migration process to a node worker so we can get non-blocking call to the route and the process will go through even if the route call dies. So far I didnt manage to get the worker pattern right so it doesnt seem to start properly.

**Review by julien51 — APPROVED** [review]
Ok let's ship to staging and do adjustments if we need!

**Pull request merged**